### ISSUE TITLE
Audit: repository learning loop from recurring defect classes

### ISSUE THREAD

**Issue by TeaShaman-cyber** [TeaShaman-cyber/theseus-needle-lab#50]
## Motivation / current evidence

A cross-product audit in `TeaShaman-cyber/theseus-repo-search-lab#10` found recurring defect classes that cut across individual files and PRs: verification-target mismatch, incomplete provenance binding, temporal/DAG phase confusion, runtime/transport boundary mistakes, and semantic overclaim.

This issue tests whether the same classes recur independently in Needle experimentation/automation rather than being specific to Repository Search.

## Intended invariant

Treat historical experiment failures, review findings, CI/dispatch/watch defects, receipt corrections, dataset/runtime contract changes, and acceptance changes as training data. Group evidence by causal class, not by file or PR, and distinguish product-specific defects from cross-product engineering/workflow patterns.

## Acceptance criteria

- Review visible Git history, issues, PR review history, CI/workflow changes, dispatch/watch paths, experiment receipts, artifact/dataset/runtime contracts, and acceptance evidence.
- Require at least two independent examples before calling something a recurring class inside this repository.
- Compare findings with `theseus-repo-search-lab#10` without assuming they must match.
- Identify which classes already have machine guards and which survive only as process knowledge.
- Record a durable checkpoint with exact issue/PR/commit/file evidence.
- No implementation in this audit; any fix becomes a narrow follow-up issue.

## Final disposition

OPEN until the first repository-wide learning checkpoint is recorded. This issue is an evidence/coordination container, not implementation or promotion authority.

**Comment by TeaShaman-cyber** [comment]
## First repo-wide learning checkpoint — Needle

Audit snapshot:

- authoritative remote `origin/main`: `39dcba38a17f3a375e7cfcdea754d671033e3e52`;
- visible Git history captured: 139 commits across current refs;
- GitHub corpus: 50 issues, 25 PRs, 132 issue comments, 45 PR review comments, 100 visible Actions runs;
- **31 top-level Codex findings** across PRs #37/#43/#45, plus extensive experiment/run postmortems in Issues #1/#4/#14/#21/#26.

The defect history independently reproduces the same causal classes found in Repository Search and Session Search, despite the domain being ML experimentation and CI orchestration rather than retrieval/indexing.

### 1. Verification illusion: successful execution repeatedly failed to prove the scientific claim

Concrete real-session examples:

- Issue #14 run `33437886183` completed successfully as infrastructure but was scientifically invalidated because training and evaluation serialized the same logical tool schema differently; schema byte order was behaviorally observable input: https://github.com/TeaShaman-cyber/theseus-needle-lab/issues/14#issuecomment-5484403125
- Stage C replicas R1/R2 initially used the same seed, so two nominal replicas were one deterministic repetition; the full run was cancelled and invalidated: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3926970566
- a cloned accepted R1 pair receipt could be renamed R2 and satisfy final acceptance until replica evidence was bound: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3932750623
- even after distinct seeds, identical/reused model artifacts across replica slots could still satisfy acceptance until artifact-set independence was enforced: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3932981413 and https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3938067085
- evaluator rows could carry stale `correct=true` inconsistent with `predicted`/`expected`, and acceptance trusted the redundant boolean: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3928236684

Invariant: **workflow success, receipt `accepted=true`, replica labels, and precomputed booleans are instruments; acceptance requires evidence that the intended independent treatment/evaluation actually occurred.**

### 2. Provenance must bind realized runtime state, not nominal labels

Independent examples:

- evaluation could consume a stale/mixed `.cact` while copying hashes from an adjacent receipt, attributing metrics to different model bytes: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3926739037
- final acceptance initially dropped model hashes, losing durable linkage to exact evaluated replicas: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3926739051
- realized seed/base checkpoint/optimizer/LoRA/token/validation configuration was initially absent from receipts: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3931696266
- Stage B launcher resource execution used `GITHUB_SHA`, which remained the **launcher** commit after checkout and was therefore written as the experiment commit in the receipt; the measurement was real but the receipt provenance was wrong: https://github.com/TeaShaman-cyber/theseus-needle-lab/issues/26#issuecomment-5518778644
- transient Actions retention also meant evaluation/training evidence could disappear even while durable notes cited it: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/43#discussion_r3931812626 and https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/43#discussion_r3932194616

Invariant: **bind the bytes/config/runtime actually consumed and preserve that binding in the final durable authority object; labels such as R1, branch name, or workflow SHA are not sufficient.**

### 3. Information loss at representation boundaries changes the experiment

Examples:

- Arm B's factorized `canonical_state` existed in sidecar data but was discarded from the actual supervised training target, so the declared treatment was not trained: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3926383474
- recovery state was computed but not fed into the curriculum, reducing the claimed adaptive mechanism to static negative resampling: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3928236705
- generic JSON canonicalization with `sort_keys=True` changed embedded tool-schema byte order; semantic JSON equality was not behavioral equality for Needle: https://github.com/TeaShaman-cyber/theseus-needle-lab/issues/26#issuecomment-5520891334
- runtime/error envelopes were collapsed into legitimate `NO_CALL`, allowing errors to count as correct negatives: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3931759301

Earlier causal experiments independently established the same principle: serialized schema order and verbalizer/token identity are behavioral state, not cosmetic formatting (Issues #14/#16/#18).

Invariant: **do not replace the realized behavioral representation with a semantically “equivalent” projection unless equivalence is experimentally established.**

### 4. Phase/DAG confusion occurs at execution and evidence boundaries

Examples:

- a manual-only workflow existing only on an experiment branch was not dispatchable until a default-branch launcher existed; repository mutation was intentionally kept distinct from execution authority (Issue #26 comment `5518664494`).
- the first launcher path failed because it assumed local `origin` existed before checkout; later the corrected launcher explicitly separated launcher SHA from checked-out experiment SHA (Issue #26 comment `5518778644`).
- PR #37 tracked curriculum artifacts remained stale after builder logic changed; a workflow `--write` would regenerate them later, but consumers of versioned data would see the invalid treatment first: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3926683889
- Stage B resource receipt semantics changed while CLI success logic still checked an old disposition literal, yielding a workflow false-red after the evidence itself had validly recorded PASS (Issue #26 comment `5521019659`).

Invariant: **execution authorization, checkout identity, generated artifacts, receipt classification, and wrapper process status are separate phases; a later repair step cannot retroactively make an earlier boundary sound.**

### 5. Evaluation coverage/independence must be proven structurally

Examples:

- heldout evaluation initially checked category counts but not exact unique case IDs, so duplicates/omissions could still pass: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3927695315
- training scopes required by the experiment were not evaluated: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3927695319
- an evaluation file could be copied into another checkpoint slot because model/checkpoint identity was not validated: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3928236691
- the curriculum hash was initially not checked before acceptance: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/37#discussion_r3932195987

This is the experimental analogue of retrieval/search coverage: **counts and green checks do not prove the intended population/surface was exercised.**

### 6. Active guidance itself became a verification surface

PR #45 shows the same class at documentation/policy level:

- candidate observations were prematurely promoted into active guidance: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/45#discussion_r3941551006
- a test checked only rule IDs, so unnumbered candidate prose could enter the active section while all tests stayed green: https://github.com/TeaShaman-cyber/theseus-needle-lab/pull/45#discussion_r3941614075

The eventual guard pins the approved active-section content rather than a weak proxy.

### Higher-order compression

Needle independently collapses to the same three meta-failures:

```text
information loss at a boundary
  canonical state / exact schema bytes / error class / artifact identity -> weaker proxy

phase confusion
  launcher vs experiment / generated vs committed artifact / evidence vs wrapper status

verification illusion
  green workflow / accepted receipt / replica count / passing test checks the wrong claim
```

### Existing guards

The repository already learned strongly from these failures: exact experiment+launcher identity, durable artifact hashes/readback, explicit replica independence, exact scope/case validation, preregistered dispositions, invalid-response scoring, behavioral-schema byte contracts, immutable receipt paths, and separation of execution success from scientific acceptance.

### Cross-product implication

These defects arose in supervised-learning experiments, replica acceptance, and CI orchestration—not in repository search or chat-history normalization. Their recurrence strongly supports a cross-product engineering/workflow pattern rather than one product's architecture.

Disposition: `FIRST LEARNING CHECKPOINT RECORDED / NO IMPLEMENTATION AUTHORIZED`.

**Comment by TeaShaman-cyber** [comment]
## Remediation follow-up created from the learning audit

The audit in #50 is now connected to an explicit implementation roadmap rather than being left as process prose:

- #51 roadmap parent
- #52 canonical QA endpoint
- #53 lifecycle/acceptance/disposition separation
- #54 resumable CI and observability contracts
- #57 backlog lifecycle reconciliation

The typed-decision research side is tracked separately in #55/#56 so process remediation does not get conflated with model-quality evidence.

GitHub Project #9 provides the coordination view; it is not authority.

No historical Needle result is rewritten by these follow-ups.
